### PR TITLE
Feature | VcfWriter | sort headers when writing to vcf

### PR DESCRIPTION
**Pull request by MayaBerg**
when using the function VcfWriter.write -> write_headers
it can take any set of headers and it will write them in the correct order. 
first:
##fileformat
last:
#CHROM POS  ID REF ALT QUAL FILTER INFO...

### CODE REVIEW

File: src/vcfio/vcf_writer.py [hunk 1]
```diff
@@ -41,10 +41,23 @@ def write_headers(self, headers: Iterable[AnyStr] = ()):
...
             headers = [REQUIRED_HEADERS[0]]
             self._write_last_header = True
-        for header in headers:
+        sorted_headers = self.sort_headers(headers)
+        for header in sorted_headers:
             self._file_descriptor.write(header)
             self._file_descriptor.write('\n')
 
+    def sort_headers(self, headers: Iterable[AnyStr] = ()):
+        def custom_sort(line):
```
Comment (Talya-dor) [inline src/vcfio/vcf_writer.py]: lets add a condition for line that is not a header. if there is a logging - lets add a warning

File: tests/test_vcf_writer.py [hunk 2]
```diff
@@ -64,3 +64,18 @@ def test_write_filetypes(self, tmp_path, output_suffix, input_file, write_header
         with VcfReader(input_file) as reader, VcfWriter(output_path) as writer:
             writer.write(headers=reader.headers, variants=reader)
         assert open_file(output_path).read(9999999).encode().strip() == Path(expected_output_file).read_bytes().strip()
+
```
Comment (Talya-dor) [inline tests/test_vcf_writer.py:65]: lets add test with line that is not a header
lets add test with sorted headers

File: src/vcfio/vcf_writer.py [hunk 3]
```diff
@@ -41,10 +41,24 @@ def write_headers(self, headers: Iterable[AnyStr] = ()):
         if not headers:
             headers = [REQUIRED_HEADERS[0]]
             self._write_last_header = True
-        for header in headers:
+        sorted_headers = self.sort_headers(headers)
```
Comment (nstern1985) [inline src/vcfio/vcf_writer.py]: could we move the custom_sort into a private member function of  the class and call it like this? 
```suggestion
        sorted_headers = sorted(headers, , key=self.__sort_headers))
```

File: src/vcfio/vcf_writer.py [hunk 4]
```diff
@@ -2,7 +2,7 @@
 
 from pathlib import Path
 from typing import TYPE_CHECKING
-
+import logging
```
Comment (nstern1985) [inline src/vcfio/vcf_writer.py]: import without a usage ?

File: src/vcfio/vcf_writer.py [hunk 5]
```diff
@@ -41,10 +41,31 @@ def write_headers(self, headers: Iterable[AnyStr] = ()):
         if not headers:
             headers = [REQUIRED_HEADERS[0]]
             self._write_last_header = True
-        for header in headers:
+        sorted_headers = self.__sort_headers(headers)
```
Comment (nstern1985) [inline src/vcfio/vcf_writer.py]: maybe change to order ?

File: src/vcfio/vcf_writer.py [hunk 6]
```diff
@@ -41,10 +41,31 @@ def write_headers(self, headers: Iterable[AnyStr] = ()):
...
+        for header in sorted_headers:
             self._file_descriptor.write(header)
             self._file_descriptor.write('\n')
 
+    def __sort_headers(self, input_headers):
+        output_headers = input_headers[:]
+        first_line_done = False
+
+        for i in range(len(input_headers)):
```
Comment (nstern1985) [inline src/vcfio/vcf_writer.py]:
```suggestion
        for i, header in enumerate(input_headers):
```

File: tests/test_vcf_writer.py [hunk 7]
```diff
@@ -64,3 +64,33 @@ def test_write_filetypes(self, tmp_path, output_suffix, input_file, write_header
...
+    @pytest.mark.parametrize("input_file, expected_output_file", (
+            ('tests/samples/sample1.vcf', 'tests/samples/sample1.info_rank.vcf'),
+    ))
+    def test_write_non_sorted_headers_at_the_end(self, input_file, expected_output_file):
```
Comment (nstern1985) [inline tests/test_vcf_writer.py]: use tmp_path for having a temporary folder

File: tests/samples/sample1.info_rank.vcf
Comment (nstern1985) [inline tests/samples/sample1.info_rank.vcf:1]: can we change the file name to be more informative like:
expected_sample_info_headers_before_last_line

File: tests/test_vcf_writer.py (hunk 7 above)
Comment (MayaBerg) [reply tests/test_vcf_writer.py:71]: look at line 72

**Review by nstern1985 — APPROVED** [review]

**Pull request merged**